src/core/lib/surface/call.c

```diff
@@ -1384,8 +1397,8 @@ static void receiving_initial_metadata_ready(grpc_exec_ctx *exec_ctx,
     }
   }
 
-  call->has_initial_md_been_received = true;
-  if (call->saved_receiving_stream_ready_bctlp != NULL) {
+  if (!gpr_atm_acq_cas(&call->recv_state, RECV_NONE,
```

Comment:
Thanks @sreecha! I didn't consider the `__ATOMIC_RELAXED` semantics in case of failure. The problem is `full_cas` also has this issue. We may need to find another solution if the semantics in failed comparison cannot be changed to `__ATOMIC_ACQUIRE`.